### ISSUE TITLE
Vanilla distance configuration (originally Nether Fortress ... issues?)

### ISSUE THREAD

**Issue by duely** [svenhjol/Charm#92]
So, I'm testing out the worldgen of a new modpack -- was trying to make sure all of the Biomes o' Plenty features had been properly disabled. However, I found a strange platform of bedrock covered in torches with horses & ocelots nearby.

![](https://i.imgur.com/YGHikJk.jpg)

![](https://i.imgur.com/58jpT44.jpg)

![](https://i.imgur.com/Q3d8STE.jpg)

![](https://i.imgur.com/NMOwHgK.jpg)

[This is the config for Charm that I've been using](https://gist.github.com/duely/42f22f70e7dae0d8a050e8c3eb2de760), with specifically the version of `Charm-1.12.2-1.1.12.jar`. [Here's a copy of the pack](https://cdn.discordapp.com/attachments/334672187728592909/588687916419776513/Lightweight-1.0.0.zip) config with override mods. Finally, [here's a copy of the world for the pack](https://drive.google.com/open?id=1dCxmvfUcYVtsg7IIPElYNgnT0nHbESwu) if you want to inspect anything.

Having looked through your code previously I can only assume either

a) that the populate function for that chunk was setting true for village generation

or b) the storage of the chunk positions is dimension agnostic and mistaking the nether for the overworld.

That explains the horses and the cats, but it doesn't really explain the bedrock?

**Comment by svenhjol** [comment]
I agree that b) is a likely culprit. In my vanilla debug environment, I tested going close to a village in the overworld so that it stored the chunks to be decorated, but not actually close enough for the decorator to trigger. Then when I went to the nether I confirmed that the chunks are still hanging around waiting to be decorated - since they're dimension agnostic then it is probable that it tried to decorate the chunks in the nether - but for me the decorator never fired when I went to nether chunk co-ords.

This is likely because the "Flower decorator" event never seems to get run in the vanilla nether. I'm guessing since BoP adds biomes to the nether, then the flower decorator is actually used there, so it picks up these undecorated chunks and starts rendering the village decorations.

The solution I'm going to use is to add the dimension ID to the village chunks so this kind of thing doesn't happen. However I have no explanation for the bedrock platform...

**Comment by svenhjol** [comment]
Attaching an alpha build with the above fix in it. I would be grateful if you could test it out in your modpack and if it's no longer causing an issue I will make it available for general release.
[Charm-1.12.2-1.1.13-alpha.zip](https://github.com/svenhjol/Charm/files/3288143/Charm-1.12.2-1.1.13-alpha.zip)

**Comment by duely** [comment]
I'll give it a go with the current configuration of the modpack, reverting the changes to BoP biomes if necessary. Specifically, I'm going to create a small mixin-mod/ASM mod to reduce the distance between villages to practically nothing, then launch a superflat world, then create a nether portal and see what happens with that -- it feels like the most efficient way to potentially pollute your chunk pool.

**Comment by svenhjol** [comment]
I actually quite like the idea of configurable village distance. Can I take a look at your ASM when you’re done building?

**Comment by duely** [comment]
Sorry, I forgot that I'd committed to get a new Arcane Archives alpha preview build out and ... well, 83 paragraphs of changelog later it's finally out. I'll try to get everything worked out and tested tomorrow -- as I'm eager to play this pack with Pyrotech!

**Comment by duely** [comment]
I had a little time to test. Using the code (I pushed the changes I made to make villages spawn more frequently to the repository linked in my previous commit) above and the current version on Forge, I found the following:

![](https://i.imgur.com/ADkH3m5.png)

The lack of villagers in the above seemed to indicate that, while I wasn't able to actually find where the entities had been added, the Charm population seemed to be happening.

Recreating the identical world and creating a Nether Portal immediately at spawn, then exploring the Nether slightly, I had the following:

![](https://i.imgur.com/Ti6SeeS.png)

I think we can safely say that the alpha version linked above fixed it. 

For the record, my ASM is extremely messy. I prefer Mixins but unfortunately I had a lot of difficulty getting them to do exactly what I wanted for this case. [This](https://github.com/duely/UndergroundVillages/blob/master/src/main/java/com/noobanidus/uv/core/transformers/SpawnTransformer.java) is the transformer that was applied, and [these](https://github.com/duely/UndergroundVillages/blob/master/src/main/java/com/noobanidus/uv/core/hooks/SpawnHooks.java) were the hooks used.

The primary changes were obviously manually setting `distance` in `MapGenVillage` to 8. The side-effect of this is that it causes everything to crash, so I also adjusted the offset that was used for random positioning to 4, and then replaced the `findNearestStructurePos` function with one that also used 4 as a step instead of 8. 

You could probably safely use reflection to find the relevant chunk generator and then modify the distance variable itself, since it is public, but obviously values must be 9 or higher. 

Additionally, I also removed the biome check in order to force villages to spawn regardless.

**Comment by svenhjol** [comment]
Great stuff. Glad the alpha works. I think I will make the village distance configurable in Charm. Will attempt reflection first, if that fails I’ll adapt your transformer code. Curious why it crashed at 8, am I missing something obvious about that?

**Comment by duely** [comment]
Oh, I should clarify further:

`MapGenVillage.java:72-73`:

```java
        k = k + random.nextInt(this.distance - 8);
        l = l + random.nextInt(this.distance - 8);```

Setting it to less than 8 without modifying the bounds as I did in my ASM will result in a value error for `Random.nextInt` talking about bounds needing to be positive. It's a slightly unclear message as it implies that you've passed a negative value -- but by "positive" it means `> 0`.

Similarly, `MapGenVillage.java:88-92`:

```java
    public BlockPos getNearestStructurePos(World worldIn, BlockPos pos, boolean findUnexplored)
    {
        this.world = worldIn;
        return findNearestStructurePosBySpacing(worldIn, this, pos, this.distance, 8, 10387312, false, 100, findUnexplored);
    }```

The `8` in this function is the `stepOffset` for the static method (which is protected, I just noticed, not private, so you would need to access it via reflection instead of the hack-job I did of using an access transformer) `findNearestStructurePosBySpacing`, which uses `distanceStep - stepOffset` as the input into `Random.nextInt` -- again, if the step is 8 or less, this will result in (by default) the same error as above.

Hence why my terrible transformer modified the hard `8` value to `4`.

With regards reflection, I think the biggest difficulty would be the fact that you have to dive through a number of layers -- first through the WorldServer to find the ChunkProvider (protected value), then through the ChunkProvider to find the ChunkGenerator (a public value, surprisingly), and then depending on the type of the generator (whether it's a `flat` world generator or an overworld generator), you'd need to either modify the MapGenVillage `villageGenerator` or, for `flat` worlds, the private map `structureGenerators` with the String key `Village`, with the private `distance` value. My confusion at the time and the reason why I used ASM was my brain being too addled to work out the right point to inject this code -- after the servers are created for each dimension but before the initial world spawn was done.

Some other considerations: 

- `size` seems to be a value that's only used by the ChunkGeneratorFlat, which is interesting as it would seem to have no default. Despite that, the value is passed in to the `Start` piece and all other pieces, so I'm not sure what its default is in that case or what kind of impact it would have on village generation.
- `minTownSeparation` exists but, as far as I can tell, has absolutely no impact and despite being initialised to 8, is never used anywhere else in the code.
- setting `distance` to a large value may result in `getNearestStructurePos` taking up more CPU time searching the map for structures than it normally would. I'm not sure what a sensible upper limit would be,

**Comment by svenhjol** [comment]
I'm going to change the title and label of this ticket for future inclusion as feature. Thanks for the detailed explanation and testing!

**Comment by duely** [comment]
Just a quick note, as this issue has changed in nature to the vanilla distance configuration, you may wish to re-open the issue (which was auto-closed with bb1c030)
